core/src/main/java/org/infinispan/commands/InvocationManager.java

```diff
@@ -0,0 +1,48 @@
+package org.infinispan.commands;
+
+import java.util.Collection;
+
+import org.infinispan.util.TimeService;
+
+/**
+ * There are three sources of invocation invalidations:
+ * 1) primary/backup writer checks expiration
+ * 2) expiration thread checks expiration
+ * 3) originator-triggered invalidation (ForgetInvocationsCommand)
+ *
+ * There can't be any concurrent 1) expirations, and 2) should occur in single thread, too
+ */
+public interface InvocationManager {
+   /**
+    * @return Time (in milliseconds) after which an {@link org.infinispan.commands.InvocationRecord} should expire.
+    */
+   long invocationTimeout();
+
+   /**
+    * @return Delegated value from {@link TimeService#wallClockTime()}
+    */
+   long wallClockTime();
```

Comment:
I didn't want to inject both `TimeService` and `InvocationManager` to all commands, as growing the commands increases GC pressure.